### PR TITLE
Robocup2019 Goalie Tactic

### PR DESCRIPTION
**Pull request by akhilveeraghanta**
### Description
Goalie Tactic to defend the net using block cone

### Testing Done

Not much testing other than in game at robocup. Could use more field testing + logic improvement, or even a better way to approach this problem. Its a good enough place holder goalie for now.

### Length Justification
N/A

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Start of document comments**: each `.cpp` and `.h` file should have a comment at the start of it. See files in the `thunderbots/software/geom` folder for examples.
- [x] **Function comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the classes defined in `thunderbots/software/geom`
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [ ] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [ ] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]
Looks like there was a weird build error with avoid areas

**Review by garethellis0 — CHANGES_REQUESTED** [review]
WHOOPS didn't submit my review a few weeks ago.......

**Comment by MathewMacDougall** [comment]
@akhilveeraghanta 	any updates?

**Comment by akhilveeraghanta** [comment]
Will have something up by thursday, sorry for the delay

**Comment by MathewMacDougall** [comment]
@akhilveeraghanta have you had a chance to update this?

**Comment by akhilveeraghanta** [comment]
Im finally done with work, will wrap this up today :crossed_fingers:

File: src/thunderbots/software/ai/hl/stp/tactic/goalie_tactic.cpp
Comment (akhilveeraghanta) [reply src/thunderbots/software/ai/hl/stp/tactic/goalie_tactic.cpp]: Ray/Line segment checks didn't really provide what we wanted as there was no way to mask a side of the rectangle when asking for an intersection, We only want intersections on the front 3 sides, with the width (the side parallel to the width of the goal) having the higher priority.

File: src/thunderbots/software/ai/hl/stp/tactic/goalie_tactic.cpp
Comment (akhilveeraghanta) [reply src/thunderbots/software/ai/hl/stp/tactic/goalie_tactic.cpp]: added

File: src/thunderbots/software/ai/hl/stp/tactic/goalie_tactic.cpp
Comment (akhilveeraghanta) [reply src/thunderbots/software/ai/hl/stp/tactic/goalie_tactic.cpp]: fixed

File: src/thunderbots/software/ai/hl/stp/tactic/goalie_tactic.h
Comment (akhilveeraghanta) [reply src/thunderbots/software/ai/hl/stp/tactic/goalie_tactic.h]: fixed

File: src/thunderbots/software/ai/hl/stp/tactic/goalie_tactic.cpp
Comment (akhilveeraghanta) [reply src/thunderbots/software/ai/hl/stp/tactic/goalie_tactic.cpp]: this can happen if the ball is behind the net, then we don't have an intersection with the goalie rectangle, in this case it defaults to the closest goal post to the ball. 

So this case will almost always be reached when the ball is at a super sharp angle or is behind the net and has no direct shot on net.

File: src/thunderbots/software/ai/hl/stp/tactic/goalie_tactic.cpp
Comment (akhilveeraghanta) [reply src/thunderbots/software/ai/hl/stp/tactic/goalie_tactic.cpp]: fixed

File: src/thunderbots/software/ai/hl/stp/tactic/goalie_tactic.cpp
Comment (akhilveeraghanta) [reply src/thunderbots/software/ai/hl/stp/tactic/goalie_tactic.cpp]: fixed

**Comment by garethellis0** [comment]
One quick comment and this should be good!

File: src/thunderbots/software/ai/hl/stp/tactic/goalie_tactic.cpp
Comment (akhilveeraghanta) [reply src/thunderbots/software/ai/hl/stp/tactic/goalie_tactic.cpp]: @garethellis0 this is expected to happen, should I put a `LOG(INFO)` here? It happens every-time there is no intersection with the ball and the net.

Comment (garethellis0) [reply src/thunderbots/software/ai/hl/stp/tactic/goalie_tactic.cpp]: Oh sorry, I left a comment here, not sure why it didn't get posted. I think just expanding the above comment to something like:
```
// if there are no intersections (ex. ball behind net), then we are out of luck. 
```
should be sufficient here

File: src/thunderbots/software/ai/hl/stp/tactic/goalie_tactic.cpp
Comment (EvanMorcom) [inline src/thunderbots/software/ai/hl/stp/tactic/goalie_tactic.cpp]: TODO

**Review by EvanMorcom — CHANGES_REQUESTED** [review]

**Comment by EvanMorcom** [comment]
Lookin' Good :fire:

File: src/thunderbots/software/ai/hl/stp/tactic/goalie_tactic.cpp
Comment (akhilveeraghanta) [reply src/thunderbots/software/ai/hl/stp/tactic/goalie_tactic.cpp]: fixed

File: src/thunderbots/software/ai/hl/stp/tactic/goalie_tactic.cpp
Comment (akhilveeraghanta) [reply src/thunderbots/software/ai/hl/stp/tactic/goalie_tactic.cpp]: The rectangle is arbitrary as we were passing in a "deflated" defense area so the goalie doesn't go to the edge of the rectangle, so its not always the `DefenseArea`

File: src/thunderbots/software/ai/hl/stp/tactic/goalie_tactic.cpp
Comment (akhilveeraghanta) [reply src/thunderbots/software/ai/hl/stp/tactic/goalie_tactic.cpp]: I'm not  sure I think that may be something @MathewMacDougall added, summoning him here

File: src/thunderbots/software/ai/hl/stp/tactic/goalie_tactic.cpp
Comment (akhilveeraghanta) [reply src/thunderbots/software/ai/hl/stp/tactic/goalie_tactic.cpp]: fixed

File: src/thunderbots/software/ai/hl/stp/tactic/goalie_tactic.cpp
Comment (akhilveeraghanta) [reply src/thunderbots/software/ai/hl/stp/tactic/goalie_tactic.cpp]: fixed

File: src/thunderbots/software/ai/hl/stp/tactic/goalie_tactic.cpp
Comment (akhilveeraghanta) [reply src/thunderbots/software/ai/hl/stp/tactic/goalie_tactic.cpp]: fixed, removed the TODO as we can't do that yet

File: src/thunderbots/software/ai/hl/stp/tactic/goalie_tactic.cpp
Comment (akhilveeraghanta) [reply src/thunderbots/software/ai/hl/stp/tactic/goalie_tactic.cpp]: There was something about the geom functions that didn't allow this to be easy and I can't remember what it was, Jonathan wrote this custom implementation which does the job pretty well, i will try to see if I can sub in some geom util functions and add more documentation w/ some ASCII art to make it more clear

File: src/thunderbots/software/ai/hl/stp/tactic/goalie_tactic.cpp
Comment (EvanMorcom) [reply src/thunderbots/software/ai/hl/stp/tactic/goalie_tactic.cpp]: OK makes sense.

File: src/thunderbots/software/ai/hl/stp/tactic/goalie_tactic.cpp
Comment (EvanMorcom) [reply src/thunderbots/software/ai/hl/stp/tactic/goalie_tactic.cpp]: I would create an issue # for this so we can remember to do it

File: src/thunderbots/software/ai/hl/stp/tactic/goalie_tactic.cpp
Comment (akhilveeraghanta) [reply src/thunderbots/software/ai/hl/stp/tactic/goalie_tactic.cpp]: added ASCII art to make it more clear

File: src/thunderbots/software/ai/hl/stp/tactic/goalie_tactic.cpp
Comment (akhilveeraghanta) [reply src/thunderbots/software/ai/hl/stp/tactic/goalie_tactic.cpp]: created a ticket and added the comment

File: src/thunderbots/software/ai/hl/stp/tactic/goalie_tactic.cpp
Comment (MathewMacDougall) [reply src/thunderbots/software/ai/hl/stp/tactic/goalie_tactic.cpp]: Ah I did this to try prevent overflow during optimization with the hungarian alrogithm. Not suuuuper great, but should be removed once we have "proper" goalie assignment, which will be changed with some Play cleanup later.

**Comment by MathewMacDougall** [comment]
Left a quick comment on the goalie code I was summoned for. Otherwise just need CI passing and LGTM

File: src/thunderbots/software/ai/hl/stp/tactic/goalie_tactic.cpp
Comment (EvanMorcom) [reply src/thunderbots/software/ai/hl/stp/tactic/goalie_tactic.cpp]: Sounds like we should create an issue# for this so we don't forget it.

Shouldn't using the std::numeric_limits prevent overflow?

File: src/thunderbots/software/ai/hl/stp/tactic/goalie_tactic.cpp
Comment (EvanMorcom) [reply src/thunderbots/software/ai/hl/stp/tactic/goalie_tactic.cpp]: If ray/rectangle intersect doesn't exist, we should at least use ray-segment in this case I think.

File: src/thunderbots/software/ai/hl/stp/tactic/goalie_tactic.cpp
Comment (MathewMacDougall) [reply src/thunderbots/software/ai/hl/stp/tactic/goalie_tactic.cpp]: Created this ticket: https://github.com/UBC-Thunderbots/Software/issues/745

I understood `numeric_limits` was the largest number representable by the system (hardware). Don't think it handles overflow, because so many different operations could cause overflow later on in different ways.

Comment (EvanMorcom) [reply src/thunderbots/software/ai/hl/stp/tactic/goalie_tactic.cpp]: If that's the case, this doesn't look very safe with the -10 then

Comment (akhilveeraghanta) [reply src/thunderbots/software/ai/hl/stp/tactic/goalie_tactic.cpp]: Added TODO and ticket near line

**Comment by MathewMacDougall** [comment]
Just need CI to pass and LGTM

**Comment by akhilveeraghanta** [comment]
Its really strange, it passes on my computer, ill have to take a closer look again tonight why those tests are failing

**Comment by MathewMacDougall** [comment]
So, uh, looks like the `goalie_tactic` tests are still failing....

I took a real quick look at the logs and the tests look geometry based. If they can't be reproduced locally maybe step through them and see how close the geometry / distance comparisons are. My only guess now is that maybe it's very small precision issues and we might need to make some equality operators a tiny bit more generous.

**Comment by akhilveeraghanta** [comment]
Yeah ill have to step through it and figure out which ones it is, try to do that by tmrw

**Comment by MathewMacDougall** [comment]
Same test failures... :cry:

**Comment by garethellis0** [comment]
@akhilveeraghanta - I'm happy to sit down and pair-program this with ya tmw if you want?

**Comment by MathewMacDougall** [comment]
Poking this again. Any luck?

**Comment by akhilveeraghanta** [comment]
The test was actually fine, there were no issues with "close enough", I just had a set of points that didn't make any sense, was looking in the wrong place.

I've made the switch to CLion, It saves so much time to change one thing and run debug again, I've been missing out. it took me 30 min to figure out what was wrong after trying it out for 2 hours with vim/gdb waiting for builds and trying to figure out the line numbers and file numbers to set breakpoints. I see the light, making the switch this year :(

**Comment by MathewMacDougall** [comment]
:sweat_smile: soooo this is going to need to be merged / rebased with the new master now that all the bazel changes are in.

**Comment by garethellis0** [comment]
@akhilveeraghanta - I can do the conflict resolution if you want, you don't deserve the pain :sweat_smile:. Just lemme know

**Comment by akhilveeraghanta** [comment]
@garethellis0 haha no worries, ill clean it up by tuesday

**Comment by jonathanlew** [comment]
@akhilveeraghanta Any updates? I need all the tactics in before #793 can proceed

**Comment by akhilveeraghanta** [comment]
rebased off of upstream master only a few conflicts, having some weird build errors with Bazel on my Mac so hoping CI will catch any bad merges

**Comment by akhilveeraghanta** [comment]
Ok build is failing, will get this up and running as soon as the Param WIP PR is up @jonathanlew plz lemme know if it starts to block you and I don't have it done by next week

**Comment by MathewMacDougall** [comment]
Still a few build issues and formatting.

**Review by garethellis0 — DISMISSED** [review]
CI is green, whoop whoop! LGTM :+1: :tada:

**Comment by garethellis0** [comment]
@EvanMorcom - this is blocked on your comments, could you please re-review?

**Review by MathewMacDougall — DISMISSED** [review]

**Comment by akhilveeraghanta** [comment]
@garethellis0 @MathewMacDougall we need a ticket to re-evaluate and redo this tactic, it was only every tested in-game and tweaked between games at robocup, once the bots are faster this will need to be revisited, we had to pull the goalie out as our bots are super slow moving side to side, but teams like ER-Force can stay back and intercept really well as their bots are hella fast (ant their intercept is really good)

This is not a well thought out tactic, it is good enough to be a placeholder goalie tactic but we need something to replace it at some point

**Comment by garethellis0** [comment]
Thanks for pointing this out @akhilveeraghanta, can you file a ticket for this as part of resolving technical debt? Probably a difficult 13 I think.

EDIT: on second thought, not sure if this should count as technical debt or not, unless it's really bad. I'll leave that up to your discretion @akhilveeraghanta

**Review by jonathanlew — CHANGES_REQUESTED** [review]
Looks good, just delete CMakelists :)

**Review by jonathanlew — DISMISSED** [review]
Good work!

**Comment by MathewMacDougall** [comment]
One more conflict...

File: src/software/ai/hl/stp/tactic/goalie_tactic.cpp
Comment (EvanMorcom) [inline src/software/ai/hl/stp/tactic/goalie_tactic.cpp]: We mention that 'block_cone_radius' is related somewhat to ROBOT_MAX_RADIUS, but the definition of the dynamic parameter doesn't include ROBOT_MAX_RADIUS. Would it be possible to change that?

File: src/software/ai/hl/stp/tactic/goalie_tactic.cpp
Comment (EvanMorcom) [inline src/software/ai/hl/stp/tactic/goalie_tactic.cpp]: I think there should be a comment here explaining what you're doing with the goalie when there is no clamped position

**Review by EvanMorcom — DISMISSED** [review]
💯 Looks good!

I just got a couple of comments.

File: src/software/ai/hl/stp/tactic/goalie_tactic.cpp
Comment (akhilveeraghanta) [reply src/software/ai/hl/stp/tactic/goalie_tactic.cpp]: updated ai.yaml

File: src/software/ai/hl/stp/tactic/goalie_tactic.cpp
Comment (akhilveeraghanta) [reply src/software/ai/hl/stp/tactic/goalie_tactic.cpp]: added a comment

**Review by MathewMacDougall — APPROVED** [review]
:pray: LGTM

**Review by jonathanlew — APPROVED** [review]
LGTM! :100:

**Comment by jsn0716** [comment]
I think `Software/src/thunderbots/software/test/ai/hl/stp/tactic/goalie_tactic.cpp` may have been misplaced after the directory changes?

**Comment by MathewMacDougall** [comment]
Yup that's being addressed in #969

**Pull request merged**
